File: andre/examples/c/module6/stack-frame-deallocation.c
//does some hack-y stuff manipulating memory on the stack
//see stack-frame-deallocation.go to see if its possible there
#include <stdio.h>
#include <stdlib.h>

const static int *global;

void a() {
	int y = 13;
}

void f() {
    char y = 'a'; // what if we put something before it?
	int x = 12;
	global = &x; //should x live on?
}

int main () {
	f();
	//function is called and f() stack is now deallocated
	printf("globals value:\t%d\nglobals ptr:\t%p\nglobals\taddress %p\n", *global, global, &global);
	a();
	//now what happens?
	printf("globals value:\t%d\n", *global);
}

File: andre/examples/go/module6/stack-frame-deallocation.go
// demonstration that a compiler should choose where var declarations exist
package main

import "fmt"

var global *int32

func f() {
	var x int32 = 12
	global = &x //x must live on

}

func g() int32 {
	var y int32 = 13
	return y
}

func main() {

	f()
	fmt.Printf("globals value:\t%d\nglobals ptr:\t%p\nglobals\taddress %p\n", *global, global, &global)
	g()
	fmt.Printf("globals value:\t%d\n", *global)
}
